File: app/javascript/src/services/api-service.js
import ErrorsHelpers from "../helpers/errors-helpers";

const DEFAULT_API_URL = '/api';

class ApiService {
  static parseArray(array, attributeName) {
    array.map(element => (
      `${attributeName}[]=${element}`
    )).join('&')
  }

  static httpResponse = async (response) => {
    const { status } = response;

    if (status === 201 || status === 204) return true;
    else if (status >= 200 && status < 400) return response.json();
    else throw response.json().then(e => new ErrorsHelpers(e));
  }

  static get = async (params) => {
    const response = await fetch(`${DEFAULT_API_URL}/${params.url}`);

    return ApiService.httpResponse(response);
  }

  static post = async (params) => {
    const response = await fetch(`${DEFAULT_API_URL}/${params.url}`, {
      body: params.body,
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'accept': 'application/json'
      }
    });

    return ApiService.httpResponse(response);
  }

  static put = async (params) => {
    const response = await fetch(`${DEFAULT_API_URL}/${params.url}`, {
      body: params.body,
      method: 'PUT',
      headers: params.headers || {
        'Content-Type': 'application/json',
        'accept': 'application/json'
      }
    });

    return ApiService.httpResponse(response);
  }

  static delete = (params) => {
    return fetch(`${DEFAULT_API_URL}/${params.url}`, {
      method: 'DELETE',
      headers: {
        'Content-Type': 'application/json',
        'accept': 'application/json'
      }
    }).then((response) => {
      if (response.status === 204) return;

      if (response.status >= 200 && response.status < 400) {
        return response.json();
      } else {
        return response.json().then(e => {throw new ErrorsHelpers(e)});
      }
    })
  }
}

export default ApiService;

File: app/services/gmail/api_service.rb
# frozen_string_literal: true

module Gmail
  class ApiService
    SCOPES = {
      gmail: 'https://www.googleapis.com/gmail/v1/',
      oauth: 'https://oauth2.googleapis.com/'
    }.freeze

    CLIENT_ID = Rails.application.credentials.google.fetch(:client_id)
    CLIENT_SECRET = Rails.application.credentials.google.fetch(:client_secret)
    REDIRECT = Rails.application.credentials.google.fetch(:redirect)

    def initialize(scope)
      @api_path = SCOPES[scope.to_sym || :gmail]
    end

    def get(path, params: {}, token:)
      url = "#{@api_path}#{path}"

      Rails.logger.info("HTTP request #{url} with params #{params.inspect}")

      HTTParty.get(
        url,
        query: params,
        headers: {
          'Authorization': "Bearer #{token}",
          'Content-Type': 'application/json'
        }
      )
    end

    def post(path, params: {}, query: {}, token:, content_type: 'application/json')
      url = "#{@api_path}#{path}"

      Rails.logger.info("HTTP request #{url} with params #{params.inspect}")

      HTTParty.post(
        url,
        query: query,
        body: params.to_json,
        headers: {
          'Authorization': "Bearer #{token}",
          'Content-Type': content_type
        }
      )
    end
  end
end
